Possible data race in the query executor

Running the test suite with the data race detector enabled reveals a possible data race:

```==================
WARNING: DATA RACE
Write at 0x00c000153858 by goroutine 105:
  sync/atomic.AddInt32()
      /usr/local/go/src/runtime/race_amd64.s:292 +0xb
  github.com/movio/bramble.(*queryExecution).executeChildStep()
      /Users/nmaquet/Workspace/bramble/query_execution.go:152 +0x6b
  github.com/movio/bramble.(*queryExecution).executeChildStep.func1()
      /Users/nmaquet/Workspace/bramble/query_execution.go:190 +0x69
  golang.org/x/sync/errgroup.(*Group).Go.func1()
      /Users/nmaquet/go/pkg/mod/golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:57 +0x94

Previous read at 0x00c000153858 by goroutine 44:
  github.com/movio/bramble.(*queryExecution).executeChildStep()
      /Users/nmaquet/Workspace/bramble/query_execution.go:153 +0x7c
  github.com/movio/bramble.(*queryExecution).executeChildStep.func1()
      /Users/nmaquet/Workspace/bramble/query_execution.go:190 +0x69
  golang.org/x/sync/errgroup.(*Group).Go.func1()
      /Users/nmaquet/go/pkg/mod/golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:57 +0x94

Goroutine 105 (running) created at:
  golang.org/x/sync/errgroup.(*Group).Go()
      /Users/nmaquet/go/pkg/mod/golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:54 +0x73
  github.com/movio/bramble.(*queryExecution).executeChildStep()
      /Users/nmaquet/Workspace/bramble/query_execution.go:189 +0x6fc
  github.com/movio/bramble.(*queryExecution).executeRootStep.func1()
      /Users/nmaquet/Workspace/bramble/query_execution.go:126 +0x69
  golang.org/x/sync/errgroup.(*Group).Go.func1()
      /Users/nmaquet/go/pkg/mod/golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:57 +0x94

Goroutine 44 (running) created at:
  golang.org/x/sync/errgroup.(*Group).Go()
      /Users/nmaquet/go/pkg/mod/golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:54 +0x73
  github.com/movio/bramble.(*queryExecution).executeChildStep()
      /Users/nmaquet/Workspace/bramble/query_execution.go:189 +0x6fc
  github.com/movio/bramble.(*queryExecution).executeRootStep.func1()
      /Users/nmaquet/Workspace/bramble/query_execution.go:126 +0x69
  golang.org/x/sync/errgroup.(*Group).Go.func1()
      /Users/nmaquet/go/pkg/mod/golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c/errgroup/errgroup.go:57 +0x94
==================
--- FAIL: TestQueryExecutionMultipleServicesWithArray (0.01s)
    testing.go:1092: race detected during execution of test
```

--- a/.github/workflows/go.yml
+++ b/.github/workflows/go.yml
@@ -15,7 +15,7 @@ jobs:
         run: go build -v ./cmd/bramble
 
       - name: Test
-        run: go test -v -coverprofile=coverage.txt -covermode=atomic $(go list ./... | grep -v examples)
+        run: go test -race -v -coverprofile=coverage.txt -covermode=atomic $(go list ./... | grep -v examples)
 
       - uses: codecov/codecov-action@v1
         name: Upload coverage

--- a/query_execution.go
+++ b/query_execution.go
@@ -9,7 +9,6 @@ import (
 	"sort"
 	"strings"
 	"sync"
-	"sync/atomic"
 
 	"github.com/vektah/gqlparser/v2/ast"
 	"github.com/vektah/gqlparser/v2/gqlerror"
@@ -35,6 +34,7 @@ type queryExecution struct {
 	graphqlClient  *GraphQLClient
 	boundaryFields BoundaryFieldsMap
 
+	mutex   sync.Mutex // protects requestCount
 	group   *errgroup.Group
 	results chan executionResult
 }
@@ -149,10 +149,13 @@ func (q *queryExecution) writeExecutionResult(step *QueryPlanStep, data interfac
 }
 
 func (q *queryExecution) executeChildStep(step *QueryPlanStep, boundaryIDs []string) error {
-	atomic.AddInt32(&q.requestCount, 1)
+	q.mutex.Lock()
+	q.requestCount += 1
 	if q.requestCount > q.maxRequest {
+		q.mutex.Unlock()
 		return fmt.Errorf("exceeded max requests of %v", q.maxRequest)
 	}
+	q.mutex.Unlock()
 
 	boundaryField, err := q.boundaryFields.Field(step.ServiceURL, step.ParentType)
 	if err != nil {

